File: frontend/src/lib/engineer-session.ts
import { apiFetch } from "@/lib/api";
import type { AttachmentMeta } from "@/lib/eworks-calculate-schema";
import {
  deleteSessionAttachment,
  getAttachmentUrl,
  uploadSessionAttachment,
} from "@/lib/eworks-session";

export type EngineerDurationType = "hourly" | "half_day" | "day_up_to_2" | "day_3_plus";

export type EngineerJobSummary = {
  quote_number: string;
  job_number: string;
  client_name: string;
  trade_name: string;
  property_address: string;
  engineer_name?: string | null;
  status: string;
};

export type EngineerSiteVisit = {
  scope?: string | null;
  site_notes?: string | null;
  findings?: string | null;
  attachments: AttachmentMeta[];
  engineer_count: number;
  labourer_count: number;
  duration_type: EngineerDurationType;
  hours?: number | string | null;
  days?: number | string | null;
  materials_required?: string | null;
  unit_cost?: number | string | null;
  parking_required: boolean;
  parking_amount?: number | string | null;
  congestion_required: boolean;
  congestion_amount?: number | string | null;
  ulez_required: boolean;
  ulez_amount?: number | string | null;
  waste_required: boolean;
  waste_amount?: number | string | null;
};

export type EngineerSession = {
  session_id: string;
  status: string;
  expires_at: string;
  job: EngineerJobSummary;
  site_visit: EngineerSiteVisit;
};

export type EngineerSiteVisitPayload = {
  scope?: string | null;
  site_notes?: string | null;
  findings?: string | null;
  engineer_count: number;
  labourer_count: number;
  duration_type: EngineerDurationType;
  hours?: number | null;
  days?: number | null;
  materials_required?: string | null;
  unit_cost?: number | null;
  parking_required: boolean;
  parking_amount?: number | null;
  congestion_required: boolean;
  congestion_amount?: number | null;
  ulez_required: boolean;
  ulez_amount?: number | null;
  waste_required: boolean;
  waste_amount?: number | null;
};

const ENGINEER_SESSION_STORAGE_PREFIX = "engineer-session:";

export function storeEngineerSessionCredentials(sessionId: string, sessionToken: string) {
  if (typeof window === "undefined") return;
  window.sessionStorage.setItem(
    `${ENGINEER_SESSION_STORAGE_PREFIX}${sessionId}`,
    sessionToken,
  );
}

export function readEngineerSessionCredentials(sessionId: string): string | null {
  if (typeof window === "undefined") return null;
  return window.sessionStorage.getItem(`${ENGINEER_SESSION_STORAGE_PREFIX}${sessionId}`);
}

export async function fetchEngineerSession(
  sessionId: string,
  sessionToken: string,
): Promise<EngineerSession> {
  const response = await apiFetch<EngineerSession>(`/api/v1/engineer/sessions/${sessionId}`, {
    headers: { "X-Session-Token": sessionToken },
  });
  return response.data;
}

export async function saveEngineerSiteVisit(
  sessionId: string,
  sessionToken: string,
  payload: EngineerSiteVisitPayload,
): Promise<{ message: string }> {
  const response = await apiFetch<{ message: string }>(
    `/api/v1/engineer/sessions/${sessionId}/site-visit`,
    {
      method: "PUT",
      body: JSON.stringify(payload),
      headers: { "X-Session-Token": sessionToken },
    },
  );
  return response.data;
}

export { uploadSessionAttachment, deleteSessionAttachment, getAttachmentUrl };

export function buildEngineerJobDetailPath(sessionId: string, sessionToken: string) {
  const params = new URLSearchParams({ token: sessionToken });
  return `/engineer/jobs/${sessionId}?${params.toString()}`;
}

File: backend/app/schemas/engineer_session.py
from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.schemas.eworks_link import SessionAttachmentMeta

EngineerDurationType = Literal["hourly", "half_day", "day_up_to_2", "day_3_plus"]

MAX_ENGINEER_NOTE_LENGTH = 8000


class EngineerJobSummary(BaseModel):
    quote_number: str
    job_number: str
    client_name: str
    trade_name: str
    property_address: str
    engineer_name: str | None = None
    status: str


class EngineerSiteVisitRead(BaseModel):
    scope: str | None = None
    site_notes: str | None = None
    findings: str | None = None
    attachments: list[SessionAttachmentMeta] = Field(default_factory=list)
    engineer_count: int = 0
    labourer_count: int = 0
    duration_type: EngineerDurationType = "hourly"
    hours: Decimal | None = None
    days: Decimal | None = None
    materials_required: str | None = None
    unit_cost: Decimal | None = None
    parking_required: bool = False
    parking_amount: Decimal | None = None
    congestion_required: bool = False
    congestion_amount: Decimal | None = None
    ulez_required: bool = False
    ulez_amount: Decimal | None = None
    waste_required: bool = False
    waste_amount: Decimal | None = None


class EngineerSessionRead(BaseModel):
    session_id: UUID
    status: str
    expires_at: datetime
    job: EngineerJobSummary
    site_visit: EngineerSiteVisitRead


class EngineerSiteVisitUpdate(BaseModel):
    scope: str | None = None
    site_notes: str | None = None
    findings: str | None = None
    engineer_count: int = Field(default=0, ge=0)
    labourer_count: int = Field(default=0, ge=0)
    duration_type: EngineerDurationType = "hourly"
    hours: Decimal | None = None
    days: Decimal | None = None
    materials_required: str | None = None
    unit_cost: Decimal | None = Field(default=None, ge=0)
    parking_required: bool = False
    parking_amount: Decimal | None = Field(default=None, ge=0)
    congestion_required: bool = False
    congestion_amount: Decimal | None = Field(default=None, ge=0)
    ulez_required: bool = False
    ulez_amount: Decimal | None = Field(default=None, ge=0)
    waste_required: bool = False
    waste_amount: Decimal | None = Field(default=None, ge=0)

    @field_validator("scope", "site_notes", "findings", "materials_required")
    @classmethod
    def validate_note_lengths(cls, value: str | None) -> str | None:
        if value is not None and len(value) > MAX_ENGINEER_NOTE_LENGTH:
            raise ValueError(f"Text must be at most {MAX_ENGINEER_NOTE_LENGTH} characters")
        return value

    def model_validate_duration(self) -> None:
        if self.duration_type == "hourly":
            if self.hours is None or self.hours <= 0:
                raise ValueError("Hours must be greater than 0 for hourly duration")
        else:
            if self.days is None or self.days <= 0:
                raise ValueError("Days must be greater than 0 for day-based duration")


class EngineerSiteVisitUpdateResponse(BaseModel):
    session_id: UUID
    status: str
    saved: bool = True
    message: str = "Site visit saved for estimator review."
